shared/json_convertible.rb

```diff
@@ -24,44 +24,56 @@ def to_object_dictionary(ignore_instance_variables: [])
         object_hash = {}
         self.instance_variables.each do |var|
           next if ignore_instance_variables.include?(var)
-          if self.class.attribute_name_to_json_proc_map.key?(var)
-            instance_variable_value = self.class.attribute_name_to_json_proc_map[var].call(self.instance_variable_get(var))
-          else
-            instance_variable_value = self.instance_variable_get(var)
-          end
-          if self.class.attribute_key_name_map.key?(var)
-            var_name = self.class.attribute_key_name_map[var]
+          if self.instance_variable_get(var).kind_of?(Array)
+            object_array = []
+            self.instance_variable_get(var).each do |obj|
+              object_array << obj.to_object_dictionary
+            end
+            var_name, = self._to_object_dictionary(var)
+            object_hash[var_name] = object_array
           else
-            var_name = var.to_s[1..-1]
+            var_name, instance_variable_value = self._to_object_dictionary(var)
+            object_hash[var_name] = instance_variable_value
           end
-          object_hash[var_name] = instance_variable_value
         end
         return object_hash
       end
+
+      protected
+
+      def _to_object_dictionary(var)
+        if self.class.attribute_name_to_json_proc_map.key?(var)
```

Comment:
I think it would be helpful to document these steps. I can imagine coming in here in the future and being confused by what it's doing at each step and why they are ordered in this way.